src/routing/bootstrap.rs

```diff
@@ -413,25 +414,34 @@ enum JoinResponse {
     },
 }
 
-// Keep receiving messages from `incoming_messages` and send them to `message_tx`.
-async fn receive_messages(
-    incoming_conns: &mut mpsc::Receiver<ConnectionEvent>,
-    mut message_tx: mpsc::Sender<(Message, SocketAddr)>,
-) {
-    while let Some(event) = incoming_conns.recv().await {
-        match event {
-            ConnectionEvent::Received(qp2p::Message::UniStream { bytes, src, .. }) => {
-                match Message::from_bytes(&bytes) {
-                    Ok(message) => {
-                        let _ = message_tx.send((message, src)).await;
+// Receiver of incoming messages that can be backed either by a raw `qp2p::ConnectionEvent` receiver
+// or by receiver of deserialized `Message` and provides a unified interface on top of them.
+enum MessageReceiver<'a> {
+    Raw(&'a mut mpsc::Receiver<ConnectionEvent>),
```

Comment:
`Receiver::recv` takes `&mut self`. Also don't want to pass it by value because we still need to use it after the bootstrap completes.

Reply:
Is `Receiver::recv` also being used within the Deserialized entry ?

or, Deserialized is only being used during bootstrap, meanwhile Raw is to be used during and after bootstrap?

Reply:
> or, Deserialized is only being used during bootstrap, meanwhile Raw is to be used during and after bootstrap

Exactly. So `Deserialized` is by value, because we don't need to use it afterwards, but `Raw` is by ref because we need to.